test_conformance/profiling/enqueueSyncFunctions.cpp

```diff
@@ -0,0 +1,286 @@
+
+    // First sync event basic monotonicity check
+    cl_ulong tQueued = 0, tSubmit = 0, tStart = 0, tEnd = 0;
+    error = clGetEventProfilingInfo(sync_event1, CL_PROFILING_COMMAND_QUEUED,
+                                    sizeof(tQueued), &tQueued, nullptr);
+    test_error(error, "Profiling sync1 QUEUED failed");
+    error = clGetEventProfilingInfo(sync_event1, CL_PROFILING_COMMAND_SUBMIT,
+                                    sizeof(tSubmit), &tSubmit, nullptr);
+    test_error(error, "Profiling sync1 SUBMIT failed");
+    error = clGetEventProfilingInfo(sync_event1, CL_PROFILING_COMMAND_START,
+                                    sizeof(tStart), &tStart, nullptr);
+    test_error(error, "Profiling sync1 START failed");
+    error = clGetEventProfilingInfo(sync_event1, CL_PROFILING_COMMAND_END,
+                                    sizeof(tEnd), &tEnd, nullptr);
+    test_error(error, "Profiling sync1 END failed");
+
+    if (check_times(tQueued, tSubmit, tStart, tEnd, device) != TEST_PASS)
+    {
+        log_error("Timestamp monotonicity failed for sync event 1\n");
+        test_status = TEST_FAIL;
+    }
+
+    cl_ulong end_set1[3] = { 0 }, start_set2[3] = { 0 };
+    for (int i = 0; i < 3; i++)
+    {
+        error =
+            clGetEventProfilingInfo(events_set1[i], CL_PROFILING_COMMAND_END,
+                                    sizeof(cl_ulong), &end_set1[i], nullptr);
+        test_error(error, "Profiling end set1 failed");
+        error =
+            clGetEventProfilingInfo(events_set2[i], CL_PROFILING_COMMAND_START,
+                                    sizeof(cl_ulong), &start_set2[i], nullptr);
+        test_error(error, "Profiling start set2 failed");
+    }
+
+    const bool is_barrier = (fn == clEnqueueBarrierWithWaitList);
+
+    log_info("Verification (%s queue, %s):\n",
+             is_out_of_order ? "out-of-order" : "in-order",
+             is_barrier ? "barrier" : "marker");
+
+    // Ordering:
+    //  - barrier: second set must start only after ALL kernels of first set end
+    //  - marker: on OOO queue ordering is not guaranteed (skip); on in-order
+    //  queue ordering is implied
+    if (is_barrier)
+    {
+        for (int i = 0; i < 3; i++)
+        {
+            if (start_set2[i] <= end_set1[0] || start_set2[i] <= end_set1[1]
+                || start_set2[i] <= end_set1[2])
+            {
+                log_error("Kernel %d in second set started too early\n", i);
+                test_status = TEST_FAIL;
+            }
+        }
+    }
+    else
+    {
+        if (is_out_of_order)
+            log_info("Marker on OOO: skipping cross-set ordering (not "
+                     "guaranteed).\n");
+        else
+            log_info(
+                "Marker on in-order: ordering implied; no explicit check.\n");
+    }
+
+    // Sync event 1 should start only after all set1 kernels finished (only
+    // meaningful on OOO)
+    if (is_out_of_order)
+    {
```

Comment:
I'd consider removing this check.  The ordering should be trivially satisfied due to the in-order queue semantics, but I think it's still woth checking, and the checks aren't very expensive.